### PR TITLE
FFI: bluechi agent flooder

### PR DESCRIPTION
**Pull request by dougsland**
Initial commit, removed a bunch of unneeded files from the original bluechi repo.

### CODE REVIEW

**Comment by dougsland** [comment]
WIP as I would like to do more cleanups and add the change to flood the controller.

**Review by Yarboa — COMMENTED** [review]
Can you please open an Issue with more detailed explanation?
IIUC, forkbomb should be achieved through requests from the contoller, is it?

File: tests/e2e/tools/FFI/bluechi-agent-flooder/build-scripts/build-push-containers.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+SCRIPT_DIR=$(dirname "$SCRIPT_DIR")/
+CONTAINER_FILE_DIR=$SCRIPT_DIR"../tests/containers/"
+
+IMAGE=""
+
+function build_and_push(){
+    buildah manifest rm $IMAGE &> /dev/null
+    buildah manifest create $IMAGE
+
+    buildah bud --tag "quay.io/bluechi/$IMAGE" \
```
[the remark this answers is not in the document]
Comment (dougsland) [reply tests/e2e/tools/FFI/bluechi-agent-flooder/build-scripts/build-push-containers.sh:14]: Hi Yariv, still need to cleanup some tools/scripts which we don't need. This is basically a bluechi repo files with agent modified.

**Comment by dougsland** [comment]
> Can you please open an Issue with more detailed explanation? IIUC, forkbomb should be achieved through requests from the contoller, is it?

Hi Yariv, this is not related to the forkbomb. It's related to a dbus attack. Basically, an attacker using a modified bluechi agent to flood the dbus communciation between controller and agent.

**Comment by dougsland** [comment]
From time to time, it floods the communication with manager sending 10000 heartbeat signals and SO FAR, everything is good, no crashes. 

```
[TEST] - Flooding controller node, sending signal 9995 to controller...
[TEST] - Flooding controller node, sending signal 9996 to controller...
[TEST] - Flooding controller node, sending signal 9997 to controller...
[TEST] - Flooding controller node, sending signal 9998 to controller...
[TEST] - Flooding controller node, sending signal 9999 to controller...
[TEST] - Flooding controller node, sending signal 10000 to controller...
```

@rhatdan @engelmi 

/cc @Yarboa @dennisbrendel @priyverm 

```
static int agent_heartbeat_timer_callback(sd_event_source *event_s
ource, UNUSED uint64_t usec, void *userdata) {
        Agent *agent = (Agent *) userdata;

        int r = 0;

        if (agent->connection_state == AGENT_CONNECTION_STATE_CONN
ECTED) {
                for (int i = 1; i <= 10000; i++) {
                        printf("[TEST] - Flooding controller node,
 sending signal %d to controller...\n", i);
                        r = sd_bus_emit_signal(
                                        agent->peer_dbus,
                                        INTERNAL_AGENT_OBJECT_PATH
,
                                        INTERNAL_AGENT_INTERFACE,
                                        "Heartbeat",
                                        "");
                        if (r < 0) {
                                bc_log_errorf("Failed to emit hear
tbeat signal: %s", strerror(-r));
                        }
                }
        }
```

@engelmi feel free to suggest more cleanups. This is just bluechi repo with some cleanups + an evil agent effort ;-)

**Comment by Yarboa** [comment]
> > Can you please open an Issue with more detailed explanation? IIUC, forkbomb should be achieved through requests from the contoller, is it?
> 
> Hi Yariv, this is not related to the forkbomb task. It's related to a dbus attack. Basically, an attacker using a modified bluechi agent to flood the dbus communciation between controller and agent.

Got it, thanks

**Comment by Yarboa** [comment]
> > > Can you please open an Issue with more detailed explanation? IIUC, forkbomb should be achieved through requests from the contoller, is it?
> > 
> > 
> > Hi Yariv, this is not related to the forkbomb task. It's related to a dbus attack. Basically, an attacker using a modified bluechi agent to flood the dbus communciation between controller and agent.
> 
> Got it, thanks

I assume it should run as an agent in QM container, is it?

**Comment by dougsland** [comment]
> > > > Can you please open an Issue with more detailed explanation? IIUC, forkbomb should be achieved through requests from the contoller, is it?
> 
> > > 
> 
> > > 
> 
> > > Hi Yariv, this is not related to the forkbomb task. It's related to a dbus attack. Basically, an attacker using a modified bluechi agent to flood the dbus communciation between controller and agent.
> 
> > 
> 
> > Got it, thanks
> 
> 
> 
> I assume it should run as an agent in QM container, is it? 
> 
> 

Correct

**Comment by engelmi** [comment]
> @engelmi This is just an initial patch to add an "evil bluechi agent" for Fredom from Interference effort. Feel free to suggest more cleanups.
In the future we will keep adding more tests in this agent.c and integrate it with CI/CD pipeline.

If I understood the use case correctly, it's basically to mimic a `bluechi-agent` and try to DoS the `bluechi-controller` right? So something like the setup as depicted below:

diagrams-Page-5

It's probably better to implement an `evil bluechi-agent` from scratch instead of copying the whole bluechi repo and changing it. When implementing the evil agent, it only needs to have a basic feature set of the agent. The programming language could also be any since its just for testing purposes - so it could be done in Python, for example. The initial TCP request might be a bit tricky to implement, but thats it. Having a dedicated agent would also simplify extending the tests in the future. 
I can try to provide a simplistic first POC for this in the next days. What do you think? @dougsland @Yarboa

**Comment by dougsland** [comment]
> > @engelmi This is just an initial patch to add an "evil bluechi agent" for Fredom from Interference effort. Feel free to suggest more cleanups.
> > In the future we will keep adding more tests in this agent.c and integrate it with CI/CD pipeline.
> 
> If I understood the use case correctly, it's basically to mimic a `bluechi-agent` and try to DoS the `bluechi-controller` right? So something like the setup as depicted below:
> 
> diagrams-Page-5
> 
> It's probably better to implement an `evil bluechi-agent` from scratch instead of copying the whole bluechi repo and changing it. When implementing the evil agent, it only needs to have a basic feature set of the agent. This has the following benefits:
> 
> * the programming language can be freely chosen (it's just for testing) - so it could be done in Python, for example
> * much simpler setup and less code for the evil agent
> * easier to extend in the future for more tests
> * clear distinction between "good" and "evil" agent
> 
> One disadvantage is, of course, the more upfront work. The initial TCP request might be a bit tricky to implement, but thats
> it.

@engelmi I like the idea.  I tried to implement something from scratch but in the end copy/past the code from the bluechi repo was faster. However, during the weekend I was thinking if would make more sense to use `#ifdef FFI` in the current code to avoid  a new agent implementation. WDYT ?
 
> I can try to provide a simplistic first POC for this in the next days. What do you think? @dougsland @Yarboa

If you believe `#ifdef FFI` is something weird for the maintenance of bluechi, yes make sense.

**Comment by engelmi** [comment]
> However, during the weekend I was thinking if would make more sense to use #ifdef FFI in the current code to avoid a new agent implementation. WDYT ?
...
If you believe #ifdef FFI is something weird for the maintenance of bluechi, yes make sense.

Adding an evil agent to BlueChi for these purposes doesn't seem like a good idea - it'll make maintenance and keeping code separated hard (as you wrote). 

I've sketched a very basic `EvilBlueChiAgent` together in this gist: https://gist.github.com/engelmi/43984268d32a535b6e74dc2b5090f7ea
From there, you can start adding the vtable to the peer bus that the agent provides and call all the methods that the manager provides.

**Comment by dougsland** [comment]
> > However, during the weekend I was thinking if would make more sense to use #ifdef FFI in the current code to avoid a new agent implementation. WDYT ?
> > ...
> > If you believe #ifdef FFI is something weird for the maintenance of bluechi, yes make sense.
> 
> Adding an evil agent to BlueChi for these purposes doesn't seem like a good idea - it'll make maintenance and keeping code separated hard (as you wrote).
> 
> I've sketched a very basic `EvilBlueChiAgent` together in this gist:
> https://gist.github.com/engelmi/43984268d32a535b6e74dc2b5090f7ea

Nice one, python is always so straightforward. 
 
> From there, you can start adding the vtable to the peer bus that the agent provides and call all the methods that the
> manager provides. Could you have a look? @dougsland

Sure thing, let's talk tomorrow!

**Comment by dougsland** [comment]
Closing in favor of: https://github.com/containers/qm/pull/242 Thanks @engelmi

**Pull request closed without merging**